modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt

```diff
@@ -111,36 +124,131 @@ class VaultWatcherService(private val tokenObserver: TokenObserver,
+                            sorting = sortByStateRefAscending()
+                        ).states
+                        newStartPos = isListInSync(newlyLoadedStatesList, specification)
+                        if (newStartPos == -1) {
+                            specification = getNewPageSpecification()
+                        }
+                    }
+                    val lastResultSetIndex = toResultSetIndex(newlyLoadedStatesList.lastIndex, specification)
+                    val shouldLoop = newlyLoadedStatesList.size == specification.pageSize
+                    return Triple(newlyLoadedStatesList.subList(newStartPos, newlyLoadedStatesList.size), lastResultSetIndex, shouldLoop)
+                }
+                private fun getOverlappingPageSpecification(startResultSetIndex: Int): PageSpecification {
+                    val specification = if (startResultSetIndex == 0) {
+                        initialSpecification
+                    }
+                    else {
+                        var numberPages = 1
+                        var dec = 1.0
+                        var pageSz = 0
+                        primes.forEach { prime ->
+                            val (intPart, decPart) = splitDouble(startResultSetIndex.toDouble() / prime)
+                            if (decPart > 0 && dec > decPart) {
+                                dec = decPart
+                                numberPages = intPart + 1     // Pages number start at 1
+                                pageSz = prime
+                            }
+                        }
+                        PageSpecification(numberPages, pageSz)
+                    }
+                    pageSpecifications.add(specification)
+                    return specification
+                }
+                private fun splitDouble(multiple: Double): Pair<Int, Double> {
+                    val intPart = floor(multiple).toInt()
+                    return Pair(intPart, multiple - intPart)
+                }
+                private fun isListInSync(states: List<StateAndRef<FungibleToken>>, specification: PageSpecification): Int {
+                    if (states.isEmpty()) {
+                        return 0
+                    }
+                    if (syncPoints.isEmpty()) {
+                        syncPoints.addFirst(SyncPoint(states.last(), specification, toResultSetIndex(states.lastIndex, specification)))
+                        return 0
+                    }
+                    val highResultSetIndex = toResultSetIndex(states.lastIndex, specification)
+                    val lowResultSetIndex = toResultSetIndex(0, specification)
+
+                    syncPoints.forEach {
+                        val listIndex = toListEntryIndex(it.lastKnownResultSetIndex, specification)
+                        if (listIndex in lowResultSetIndex..highResultSetIndex && states[listIndex] == it.stateAndRef) {
```

Comment:
Oops. Now fixed by just checking the index is in range. I would like to keep this if in. As hopefully it would be the normal case. i.e. when nothing has changed in the list we don't do any extra searching etc. So performance wise should be the same as now.